config.md

```diff
@@ -25,19 +25,61 @@ Changing it means creating a new derived image, instead of changing the existing
 
 ### Layer DiffID
 
-A layer DiffID is a SHA256 digest over the layer's uncompressed tar archive and serialized in the descriptor digest format, e.g., `sha256:a9561eb1b190625c9adb5a9513e72c4dedafc1cb2d4c5236c9a6957ec7dfd5a9`.
+A layer DiffID is the digest over the layer's uncompressed tar archive and serialized in the descriptor digest format, e.g., `sha256:a9561eb1b190625c9adb5a9513e72c4dedafc1cb2d4c5236c9a6957ec7dfd5a9`.
 Layers must be packed and unpacked reproducibly to avoid changing the layer DiffID, for example by using tar-split to save the tar headers.
 
 NOTE: Do not confuse DiffIDs with [layer digests](manifest.md#image-manifest-property-descriptions), often referenced in the manifest, which are digests over compressed or uncompressed content.
 
 ### Layer ChainID
 
 For convenience, it is sometimes useful to refer to a stack of layers with a single identifier.
-This is called a `ChainID`.
-For a single layer (or the layer at the bottom of a stack), the
-`ChainID` is equal to the layer's `DiffID`.
-Otherwise the `ChainID` is given by the formula:
-`ChainID(layerN) = SHA256hex(ChainID(layerN-1) + " " + DiffID(layerN))`.
+While a layer's `DiffID` identifies a single changeset, the `ChainID` identifies the subsequent application of those changesets.
+This ensures that we have handles referring to both the layer itself, as well as the result of the application of a series of changesets.
+Use in combination with `rootfs.diff_ids` while applying layers to a root filesystem to uniquely and safely identify the result.
+
+#### Definition
+
+The `ChainID` of an applied set of layers is defined with the following recursion:
+
+```
+ChainID(L₀) =  DiffID(L₀)
+ChainID(L₀|...|Lₙ₋₁|Lₙ) = Digest(ChainID(L₀|...|Lₙ₋₁) + " " + DiffID(Lₙ))
```

Comment:
> @wking this is defining the mathematical formula for a digest and reasoning behind it.

Agreed.

> Defining the algorithm or making requirements about algorithms matching is unrelated to the definition of a chain ID.

I agree that defining SHA-256 is unrelated, but I don't agree that algorithm selection is unrelated to the chainID definition.

> This is a perfectly valid chain id as the properties are reserved.
>
>     Digest("sha256", "sha512:edf2182426048661521fe12414e289f2187c9cb6df93bd05c17bb424b7c40804 sha256:bf07a7fbb825fc0aae7bf4a1177b2b31fcf8a3feeaf7092761e18c859ee52a9c")

I'm unclear on what you mean by “the properties are reserved”.

> The processing creating the chain ID is also responsible for providing the diff IDs…

This is not necessarily true.  DiffIDs are part of the distributed config, so they are created by the config author.  ChainIDs may be created by the config author (in which case your statement is true), but they may also be created by the config consumer (e.g. for identifying directories with unpacked/applied stacks of layers) and in that case your statement is false.

> …the algorithms of those diff digests are completely irrelevant to the chain id computation and the algorithm used for the end chain ID does not need to match the diff IDs.

Agreed.

> The chain ID algorithm can be set by the implementation of the chain ID and as with all cases where a digest is used, no algorithm should be preferred by specification.

That sounds like an argument for the [`ChainID(algo, L₀|...|Lₙ₋₁|Lₙ)` phrasing I recommended][1].  Without that, there is no way to consistently check a claimed `chainID`.  Using my example values, it's not clear how, you could verify that a `ChainID(L₀|L₁|L₂)` of  `sha256:d58dac7b666551a5ddaa5598e792035607a2611f29ad37082881a5bd1f462f49` was accurate unless you guessed (somehow) that the creator had used `sha512` for the `ChainID(L₀|L₁)` digest.

Leaving the algorithm choice unrestricted (and thus allowing algorithm swaps), means that there is no guess-free way of validating `ChainID(L₀|…|Lₙ)` for `n > 1`.  You *could* define `ChainID` that way, but when `ChainID(algo, L₀|...|Lₙ₋₁|Lₙ)` avoids that limitation so easily, I don't see why we wouldn't want to use it.

[1]: https://github.com/opencontainers/image-spec/pull/586#discussion_r103773769